### PR TITLE
Check for option duplicates

### PR DESCRIPTION
**Pull request by sfc-gh-jsikorski**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added new automated tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
* added `--options-structure` option, that returns tree of all commands with possible option
* added test, to check if no options are duplicated - as we have duplicates, it is currently marked as `skip`, until we will decide, how to resolve conflicts

### CODE REVIEW

File: src/snowcli/app/dev/commands_structure.py [hunk 1]
```diff
@@ -11,12 +12,27 @@ class _Node:
...
 
     def print(self):
         print("    " * self.level, self.name)
         for ch in self.children.values():
             ch.print()
 
+    def print_with_options(self):
+        options = self._prepare_options_dict()
+        pretty = json.dumps(options, indent=4)
+        print(pretty)
```
Comment (sfc-gh-turbaszek) [inline src/snowcli/app/dev/commands_structure.py:25]: Can you show example result of this callback?

File: tests/testing_utils/result_assertions.py [hunk 2]
```diff
@@ -8,3 +10,36 @@ def assert_that_result_is_usage_error(
...
+        return duplicates
+
+    return None
+
+
+def check_options_for_duplicates(options: Dict[str, List[str]]) -> List[str]:
+    RESERVED_FLAGS = ["-h", "--help"]  # noqa: N806
+    flags = [flag for option in options.values() for flag in option]
+    return list(
+        set(
```
Comment (sfc-gh-turbaszek) [inline tests/testing_utils/result_assertions.py:38]: Do we need to cast set to list?

File: src/snowcli/app/dev/commands_structure.py (hunk 1 above)
Comment (sfc-gh-jsikorski) [reply src/snowcli/app/dev/commands_structure.py:25]: It`s pretty huge, but for one command and global options, it looks like:
`{
    "options": {
        "version": [
            "--version"
        ],
        "docs": [
            "--docs"
        ],
        "structure": [
            "--structure"
        ],
        "options_structure": [
            "--options-structure"
        ],
        "info": [
            "--info"
        ],
        "configuration_file": [
            "--config-file"
        ],
        "pycharm_debug_library_path": [
            "--pycharm-debug-library-path"
        ],
        "pycharm_debug_server_host": [
            "--pycharm-debug-server-host"
        ],
        "pycharm_debug_server_port": [
            "--pycharm-debug-server-port"
        ],
        "disable_external_command_plugins": [
            "--disable-external-command-plugins"
        ],
        "commands_registration": [
            "--commands-registration"
        ],
        "install_completion": [
            "--install-completion"
        ],
        "show_completion": [
            "--show-completion"
        ]
    },
    "connection": {
        "options": {
            "install_completion": [
                "--install-completion"
            ],
            "show_completion": [
                "--show-completion"
            ]
        },
        "list": {
            "options": {
                "format": [
                    "--format"
                ],
                "verbose": [
                    "--verbose",
                    "-v"
                ],
                "debug": [
                    "--debug"
                ]
            }
        },
        "add": {
            "options": {
                "connection_name": [
                    "--connection-name",
                    "-n"
                ],
                "account": [
                    "--account",
                    "-a",
                    "--accountname"
                ],
                "user": [
                    "--user",
                    "-u",
                    "--username"
                ],
                "password": [
                    "--password",
                    "-p"
                ],
                "role": [
                    "--role",
                    "-r"
                ],
                "warehouse": [
                    "--warehouse",
                    "-w"
                ],
                "database": [
                    "--database",
                    "-d"
                ],
                "schema": [
                    "--schema",
                    "-s"
                ],
                "host": [
                    "--host",
                    "-h"
                ],
                "port": [
                    "--port",
                    "-P"
                ],
                "region": [
                    "--region",
                    "-R"
                ],
                "format": [
                    "--format"
                ],
                "verbose": [
                    "--verbose",
                    "-v"
                ],
                "debug": [
                    "--debug"
                ]
            }
        },
        "test": {
            "options": {
                "connection": [
                    "--connection",
                    "-c",
                    "--environment"
                ],
                "format": [
                    "--format"
                ],
                "verbose": [
                    "--verbose",
                    "-v"
                ],
                "debug": [
                    "--debug"
                ]
            }
        }
    }
}   `

File: tests/testing_utils/result_assertions.py (hunk 2 above)
Comment (sfc-gh-jsikorski) [reply tests/testing_utils/result_assertions.py]: Good point, changed

File: src/snowcli/app/dev/commands_structure.py (hunk 1 above)
Comment (sfc-gh-turbaszek) [reply src/snowcli/app/dev/commands_structure.py:25]: Do we need it as command callbck? The other callback we have have some use-cases (like copying to doc), But I'm not sure about this one

Comment (sfc-gh-jsikorski) [reply src/snowcli/app/dev/commands_structure.py:25]: It seemed reasonable, to place it there - with command structure, version etc.
But i`m open to suggestions :)

File: src/snowcli/app/dev/commands_structure.py [hunk 3]
```diff
@@ -26,9 +42,18 @@ def generate_commands_structure(command: Command, root: _Node | None = None):
...
     if hasattr(command, "commands"):
         for command_name, command_info in command.commands.items():
             if command_name not in root.children:
                 root.children[command_name] = _Node(command_name, level=root.level + 1)
             generate_commands_structure(command_info, root.children[command_name])
     return root
+
+
+def generate_options_structure(command: Command, root: Optional[_Node] = None):
+    pass
```
Comment (sfc-gh-turbaszek) [inline src/snowcli/app/dev/commands_structure.py:59]: Is it necessary?

File: tests/test_cli.py [hunk 4]
```diff
@@ -1,6 +1,9 @@
+import json
+
 import pytest
 
 from tests.testing_utils.fixtures import *
+from tests.testing_utils.result_assertions import find_conflicts_in_options_dict
```
Comment (sfc-gh-turbaszek) [inline tests/test_cli.py:6]: I would keep it here, in this file. This util won't be uses probably anywhere else than this tests, right?

Comment (sfc-gh-jsikorski) [reply tests/test_cli.py]: Done

File: src/snowcli/app/dev/commands_structure.py (hunk 3 above)
Comment (sfc-gh-jsikorski) [reply src/snowcli/app/dev/commands_structure.py]: Not really 😄

File: src/snowcli/app/cli_app.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import json
```
Comment (sfc-gh-pjob) [inline src/snowcli/app/cli_app.py:3]: unused import?

File: src/snowcli/app/cli_app.py [hunk 6]
```diff
@@ -23,8 +24,8 @@
 from snowcli.app.main_typer import SnowCliMainTyper
 from snowcli.config import config_init, cli_config
 from snowcli.output.formats import OutputFormat
-from snowcli.output.printing import print_result
-from snowcli.output.types import CollectionResult
+from snowcli.output.printing import print_result, print_structured
```
Comment (sfc-gh-pjob) [inline src/snowcli/app/cli_app.py:27]: unused imports?

File: src/snowcli/app/cli_app.py [hunk 7]
```diff
@@ -88,6 +89,16 @@ def _version_callback(value: bool):
         raise typer.Exit()
 
 
+@_do_not_execute_on_completion
+@_commands_registration.after
+def _options_structure_callback(value: bool):
+    if value:
+        ctx = click.get_current_context()
+        output_json = generate_commands_structure(ctx.command).print_with_options()
```
Comment (sfc-gh-pjob) [inline src/snowcli/app/cli_app.py:97]: output_json variable is unused, IMO you don't need it

File: src/snowcli/app/dev/commands_structure.py (hunk 1 above)
Comment (sfc-gh-pjob) [reply src/snowcli/app/dev/commands_structure.py:25]: Do we plan to use it other way than in tests?

Maybe we can call `generate_commands_structure` directly from tests?

Comment (sfc-gh-pjob) [reply src/snowcli/app/dev/commands_structure.py:25]: WDYT?

File: tests/test_cli.py [hunk 8]
```diff
@@ -25,3 +28,39 @@ def test_namespace(namespace, expected, runner):
...
+    keys = list(options_dict.keys())
+    duplicates = {}
+    if "options" in keys:
+        if opts_duplicates := check_options_for_duplicates(options_dict["options"]):
+            duplicates[path] = opts_duplicates
+        keys.remove("options")
+
+    for key in keys:
+        if key_duplicates := find_conflicts_in_options_dict(key, options_dict[key]):
+            duplicates[key] = key_duplicates
```
Comment (sfc-gh-pjob) [inline tests/test_cli.py:53]:
```suggestion
        if key_duplicates := find_conflicts_in_options_dict(f"{path} {key}", options_dict[key]):
            duplicates.update(key_duplicates)
```

I think it produces more readable output.

<details>
<summary>Your output (click to expand)</summary>

```
{
  'connection': {
    'add': {
      'add': {
        '-h'
      }
    }
  },
  'snowpark': {
    'function': {
      'create': {
        'create': {
          '-h'
        }
      },
      'update': {
        'update': {
          '-h',
          '-a'
        }
      }
    },
    'jobs': {
      'create': {
        'create': {
          '-c'
        }
      },
      'logs': {
        'logs': {
          '-c'
        }
      }
    },
    'procedure': {
      'create': {
        'create': {
          '-h'
        }
      },
      'describe': {
        'describe': {
          '-p'
        }
      },
      'drop': {
        'drop': {
          '-p'
        }
      },
      'execute': {
        'execute': {
          '-p'
        }
      },
      'update': {
        'update': {
          '-h'
        }
      }
    },
    'services': {
      'create': {
        'create': {
          '-p'
        }
      }
    }
  }
}
```

</details>

<details>
<summary>My output (click to expand)</summary>

```
{
  'snow connection add': {
    '-h'
  },
  'snow snowpark function create': {
    '-h'
  },
  'snow snowpark function update': {
    '-a',
    '-h'
  },
  'snow snowpark jobs create': {
    '-c'
  },
  'snow snowpark jobs logs': {
    '-c'
  },
  'snow snowpark procedure create': {
    '-h'
  },
  'snow snowpark procedure describe': {
    '-p'
  },
  'snow snowpark procedure drop': {
    '-p'
  },
  'snow snowpark procedure execute': {
    '-p'
  },
  'snow snowpark procedure update': {
    '-h'
  },
  'snow snowpark services create': {
    '-p'
  }
}
```

</details>

Comment (sfc-gh-jsikorski) [reply tests/test_cli.py]: Good point. Done

File: src/snowcli/app/cli_app.py (hunk 7 above)
Comment (sfc-gh-jsikorski) [reply src/snowcli/app/cli_app.py]: Done

File: src/snowcli/app/cli_app.py (hunk 6 above)
Comment (sfc-gh-jsikorski) [reply src/snowcli/app/cli_app.py]: Deleted

File: src/snowcli/app/cli_app.py (hunk 5 above)
Comment (sfc-gh-jsikorski) [reply src/snowcli/app/cli_app.py]: Deleted

**Pull request closed without merging**